core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -624,96 +635,259 @@ public void addPhysicallyWrittenRows(long rows) {
+                            .$(", from=").$(path).$(", to=").$(detachedPath).I$();
                 }
             }
             path.trimTo(rootLen);
-            other.trimTo(rootLen);
+            detachedPath.trimTo(detachedRootLen);
         }
+    }
+
+    public StatusCode detachPartition(long timestamp) {
+        assert metadata.getTimestampIndex() > -1;
 
-        return StatusCode.OK;
+        if (!PartitionBy.isPartitioned(partitionBy)) {
+            return StatusCode.TABLE_NOT_PARTITIONED;
+        }
+
+
+        int partitionIndex = txWriter.getPartitionIndex(timestamp);
+        if (partitionIndex == -1) {
+            assert !txWriter.attachedPartitionsContains(timestamp);
+            return StatusCode.PARTITION_ALREADY_DETACHED;
+        }
+
+        long maxTimestamp = txWriter.getMaxTimestamp();
+        if (timestamp == getPartitionLo(maxTimestamp)) {
+            return StatusCode.PARTITION_IS_ACTIVE;
+        }
+        long minTimestamp = txWriter.getMinTimestamp();
+
+        if (inTransaction()) {
+            LOG.info()
+                    .$("committing open transaction before applying detach partition command [table=")
+                    .$(tableName)
+                    .$(", partition=")
+                    .$ts(timestamp)
+                    .I$();
+            commit(CommitMode.SYNC, 0);
+        }
+
+        try {
+            // path: partition folder to be detached
+            long partitionNameTxn = txWriter.getPartitionNameTxn(partitionIndex);
+            setPathForPartition(path, rootLen, partitionBy, timestamp, partitionNameTxn);
+            if (!ff.exists(path.$())) {
+                LOG.error()
+                        .$("partition folder does not exist [path=")
+                        .$(path)
+                        .$(']')
+                        .$();
+                return StatusCode.PARTITION_FOLDER_DOES_NOT_EXIST;
+            }
+
+            // detachedPath: detached partition folder
+            if (!ff.exists(detachedPath.trimTo(detachedRootLen).slash$())) {
+                // the detached and standard folders can have different roots
+                // (server.conf: cairo.sql.detached.root)
+                if (0 != ff.mkdirs(detachedPath, mkDirMode)) {
+                    LOG.error()
+                            .$("cannot create detached partition folder [path=")
+                            .$(detachedPath)
+                            .$(']')
+                            .$();
+                    return StatusCode.PARTITION_DETACHED_FOLDER_CANNOT_CREATE;
+                }
+            }
+            setPathForPartition(detachedPath.trimTo(detachedRootLen), partitionBy, timestamp, false);
+            detachedPath.put(DETACHED_DIR_MARKER).$();
+            final int detachedPathLen = detachedPath.length();
+            if (ff.exists(detachedPath)) {
+                LOG.error()
```

Comment:
minor: this perfectly fits one line, making this method super long is not great, can you keep logging concise where possible?